### PR TITLE
docs(mvux): MVUX Overview rewrite

### PR DESCRIPTION
**Pull request by weitzhandler**
## PR Type
- Documentation content changes

## PR Checklist

Please check if your PR fulfills the following requirements:
- [x] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [x] Contains **NO** breaking changes

### CODE REVIEW

File: doc/Overview/Mvux/WIP_Overview.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+That should seem simple enough. All it needs to do is put a number on the screen.
+Although this seems like an easy problem, as is often the case, there are more details to consider than may be immediately apparent.
+
+- What if the external data isn't immediately available when starting the app?
+- How to show that data is being initially loaded? Or updated?
+- What if no data is available?
+- What if there's an error in obtaining or processing the data?
+- How to keep the app responsive while requesting or updating the UI?
+- How to keep track of the selected item or items?
```
Comment (mrlacey) [inline doc/Overview/Mvux/WIP_Overview.md:21]: It's not obvious how this relates to the app described above. What is there to select?

File: doc/Overview/Mvux/WIP_Overview.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+- Does the app need to periodically request new data or listen to the external source provide it?
+- How do we avoid threading or concurrency issues when handling new data in the background?
+- How do we make sure the code is testable?
+
+Individually, these questions and scenarios are simple to handle, but hopefully, they highlight that there is more to consider in even a very trivial application. Now imagine an application that you need to build, and with more complex data and UIs, the potential for complexity and the amount of required code can grow enormously.
+
+MVUX is a response to such situations and makes it easier to handle the above scenarios.  
+We'll look at the pattern and how to use it to make this app. 
+
+## What is MVUX?
```
Comment (mrlacey) [inline doc/Overview/Mvux/WIP_Overview.md:33]: This opening should probably also make reference to the MVU pattern in some way. It's absence is noticeable.

File: doc/Overview/Mvux/WIP_Overview.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+        DataContext = new BindableWeatherModel(new WeatherService());
+    }
+}
+```
+
+Within the template, the `Data` property is used to access the data that was obtained from the service, currently available by the IFeed. In this case it's the most recent `WeatherInfo` result from the service.
+The `Data` property is bound to the `DataContext` of the `TextBlock`, making it possible to bind the `Text` property of the `TextBlock` to the `Temperature` property.
+
+The `Refresh` command is exposed by the `FeedView` and will cause the Feed to requery its source (the service).
```
Comment (mrlacey) [inline doc/Overview/Mvux/WIP_Overview.md:176]: From looking at the XAML above, the `Refresh` command is on the `Button`, not the FeedView. The actual command looks like it is something "exposed" by the generated/bindable wrapper of the model.  Where does this actually exist? Why does the reader need to know? What does this have to do with the app initially described above, which had no mention of a refresh button? The description said it was going to periodically check for or request new values.

File: doc/Overview/Mvux/WIP_Overview.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+> [!NOTE]
+> For the full example see [How to create a feed](xref:Overview.Mvux.HowToSimpleFeed)
```
Comment (mrlacey) [inline doc/Overview/Mvux/WIP_Overview.md:179]: For an example of "Using MVUX to create an app" we haven't seen how to create an app, and we haven't seen much of the functionality described in the trivial app above.
Many of the benefits alluded to in the opening weren't covered in the example, and it's not clear what the code that is shown does in terms of bringing benefits to the reader.

**Review by mrlacey — CHANGES_REQUESTED** [review]
It could also do with being passed through a grammar checker. (Grammar issues not flagged by me.)

File: doc/Overview/Mvux/WIP_Overview.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+## What is MVUX?
+
+It stands for **M**odel, **V**iew, **U**pdate, e**X**tended.
+
+Looking at each individual element, it's easiest to start with the View.
+
+### View
+
+The **View** is the UI. You can write it with XAML, C#, or a combination of the two, much as you would when using another design pattern. You would use data binding to bind to the ViewModel and allow separating the View from ViewModel. MVUX does use a ViewModel, but in a different way than the MVVM pattern. For now let's refer to the ViewModel as just Model.
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:41]: We are going to have to decide on the terminology here. Are we using ViewModel to refer to the BPM and use Model instead of Model? Or we're going to refer to Model as just ViewModel and keep calling the BPM BPM?
I think we should just explain that Model in MVUX is in place of ViewModel in MVVM and keep calling it Model. The BRP should remain BRP, or something else NOT ViewModel.
After terminology is agreed on I'll change them throughout the document.
@mrlacey @nickrandolph @dr1rrb

File: doc/Overview/Mvux/WIP_Overview.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+### Update
+
+When the user interacts with the View and provides input or action, it will trigger an **Update** that changes the Model.
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:45]: > Especially as the example app described above does not include any user interaction with the View.

There's a Refresh button in the app, that's the interaction.

> What's the difference between "input" and "action"? Will the reader understand the difference and why you make this separation?

I was referring to input as text and values vs interaction as button/command, but if that's not clear I'm going to remove it and just use 'user input' here.

File: doc/Overview/Mvux/WIP_Overview.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+In the standard MVU pattern, an updated Model will create a new View for the Model. In MVUX however, the existing View will be updated via data-binding by a ViewModel generated by MVUX, that wraps around the Model class, and tracks changes and differences to the Model.
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:47]:
```suggestion
In the standard MVU pattern, an updated Model will create a new View for the Model.  
In MVUX the existing View will be updated via data-binding by a ViewModel generated by MVUX, the ViewModel wraps around the Model class, and tracks changes to the Model.
```

[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:47]: > "data-binding" has a hyphen here but not elsewhere. It should be consistent throughout the document.

And what should be the standard with or without a hyphen?
@nickrandolph

File: doc/Overview/Mvux/WIP_Overview.md [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+The change detection is similar to using the `INotifyPropertyChanged` interface, except MVUX automatically handles the comparison and change detection for you, using the generated ViewModel.
+
+### Model
+
+The **Model** that the Update creates (and is wrapped in a generated ViewModel) is typically a simple class or record that contains only data and no logic. For this reason, a record is usually the most appropriate data structure for the Model. C# records are immutable (read-only) and the use of them are encouraged by MVUX. It will be explained further later.
+
+Working with immutable data structures can be a challenging transition for developers who have only previously only worked with `INotifyPropertyChanged` implementing classes. The fundamental difference is that since a record is immutable, making a change to it means creating a new object with different data instead of modifying data within an existing one. Don't worry, the "X" part of the MVUX implementation takes care of this for you.
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:54]: We didn't say there are many. There are definitely some. And saying this sentence calls for any dev even those who did use rx etc to absorb the new info by preparing focus.

[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:54]: Feels a bit verbose, but is clearer IMO. I'd even use 'record' again instead of 'it'.

File: doc/Overview/Mvux/WIP_Overview.md [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+### Extended
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:56]:
```suggestion
### eXtended
```

File: doc/Overview/Mvux/WIP_Overview.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+Now that you know the fundamentals of the pattern, let's see how to use it to create the simple weather app mentioned above. 
+
+## Using MVUX to create an app
+
+This example creates an `IFeed<WeatherInfo>` that will return the weather, which includes the temperature, loaded asynchronously from a weather service.
+
+### Service
+
+The `WeatherInfo` entity wraps the temperature, and in a real application would be extended to hold other weather information such as humidity or rainfall:
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:84]: > Why is this record called an "entity" here?

It's a data entity received from the service.

> "wrapping the temperature" is not a way I've heard "having a property (or value)" described before. Is there a reason for this phrasing?

```suggestion
The `WeatherInfo` entity received from the service holds the temperature.
```

File: doc/Overview/Mvux/WIP_Overview.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+```c#
+public interface IWeatherService
+{
+    ValueTask<WeatherInfo> GetCurrentWeather(CancellationToken ct);
+}
+```
+
+This is the implementation of the `IWeatherServiec` interface.
+It returns a random value as the current weather, after a delay of one second, to simulate the asynchronous interaction with a remote server:
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:100]:
```suggestion
It returns a random value as the current weather (in Fahrenheit degrees), after a delay of one second, to simulate the asynchronous interaction with a remote server:
```

File: doc/Overview/Mvux/WIP_Overview.md [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+
+```c#
+public partial record WeatherModel(IWeatherService WeatherService)
+{
+    public IFeed<WeatherInfo> CurrentWeather => Feed.Async(WeatherService.GetCurrentWeather);
+}
+```
+
+> [!IMPORTANT]  
+> For the code generation to work, mark the Models and entities with the `partial` modifier, and have the Feed properties' access modifier as `public`.
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:126]: > What are the "Models and entities"?

Entities are commonly used in C# to define data objects. I'm happy to add a description on the first occurrence of entity about this.

> How does the code generation know which are the Models?

By the 'Model' suffix

File: doc/Overview/Mvux/WIP_Overview.md [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+{
+    public IFeed<WeatherInfo> CurrentWeather => Feed.Async(WeatherService.GetCurrentWeather);
+}
+```
+
+> [!IMPORTANT]  
+> For the code generation to work, mark the Models and entities with the `partial` modifier, and have the Feed properties' access modifier as `public`.  
+You can learn more about partial classes and methods in [this article](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/partial-classes-and-methods).
+
+The `CurrentWeather` property is an `IFeed<WeatherInfo>`. It uses as an asynchronous gateway to the service, and also holds information about the current status of the data.
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:129]: I tried to be concise here as to not distract the reader.

Does 'asynchronous interaction point with the service' sound any less 'waffly'? 'querying point'? Any other suggestions?

File: doc/Overview/Mvux/WIP_Overview.md [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+The MVUX code-generator (the X in MVUX) generates a data-binding-ready clone of the Model (prefixed `Bindable`), it re-generates its properties, and also introduces commands for all of its public methods that meet a certain criteria, which will be discussed later.
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:131]: > "The MVUX code-generator (the X in MVUX)" all of the "X" or part of it? - Above, it was said that the X was multiple things; this implies one.

```suggestion
The MVUX code-generator (part of the X in MVUX) generates a data-binding-ready clone of the Model (prefixed `Bindable`), it re-generates its properties, and also introduces commands for all of its public methods that meet a certain criteria, which will be discussed later.
```

[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:131]: > Do we see the bindable clone of the model anywhere? Or could we if we wanted to?

Of course. There's a dedicated page about how to inspect the generated code.

File: doc/Overview/Mvux/WIP_Overview.md [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+### View
+
+A special control, the `FeedView` is used to display data exposed as an `IFeed<T>`, and provides different display templates for the various statuses of the Feed (e.g. loading, refreshing, error etc.).
+
+The `FeedView` is data-bound to the `CurrentWeather` property which aligns with the `CurrentWeather` property exposed on the `WeatherModel`:
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:137]: Yes, the property in the BRP acts as a data-binding-ready property that interacts with the Model's `CurrentWeather` feed property.

I'll rephrase it to sharpen it out.

File: doc/Overview/Mvux/WIP_Overview.md [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+<Page x:Class="WeatherApp.MainPage"
+	  xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+	  xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+	  xmlns:mvux="using:Uno.Extensions.Reactive.UI">
+
+	<mvux:FeedView Source="{Binding CurrentWeather}">
+		<DataTemplate>
+			<StackPanel>
+				<TextBlock DataContext="{Binding Data}"
+						   Text="{Binding Temperature}" />
```
Comment (weitzhandler) [inline doc/Overview/Mvux/WIP_Overview.md:149]:
```suggestion
				<TextBlock Text="{Binding Data.Temperature}" />
```

File: doc/Overview/Mvux/WIP_Overview.md [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+```xml
+<Page x:Class="WeatherApp.MainPage"
+	  xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+	  xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+	  xmlns:mvux="using:Uno.Extensions.Reactive.UI">
+
+	<mvux:FeedView Source="{Binding CurrentWeather}">
+		<DataTemplate>
+			<StackPanel>
+				<TextBlock DataContext="{Binding Data}"
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md:148]: The reason I did it is to emphasize that there is a `Data` property here, nevermind I'll remove it.

File: doc/Overview/Mvux/WIP_Overview.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (dr1rrb) [reply doc/Overview/Mvux/WIP_Overview.md]: @weitzhandler what are "BPM" and "BRP"?

IMHO, we should not use the term "ViewModel" as it refers to the MVVM pattern.

[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md]: @dr1rrb Bindable Proxy Model. BRP my mistake, meant BPM.

[the remark this answers is not in the document]
Comment (dr1rrb) [reply doc/Overview/Mvux/WIP_Overview.md]: Great that's what I was expecting :) 

So to complete my answer, currently in the code the terms I tried to use are "entity" / "model" / "<bindable|generated> proxy" / "view", and I maintain that we should use "view model" at all as you are suggesting.

However my really concern is about the naming of the "entity" and I think that's why we are so confused: for me it's the "model" in MVVM and MVU. And what we are naming the "model" is the "View model" in the MVVM but has not real explicit equivalent in MVU ("update"?).

Maybe we would like to change that names to "model" / "<TBD - but NOT ViewModel>" / "BPM" / "view"

The part for which we should then determine a name is the place where we receive commands and we generate an updated model, even if I don't like to use a verb for that, it seems that the term used in MVU is "update".

To summurize:

| MVVM | MVU | Current | Proposal |
| --- | --- | --- | --- |
| Model | Model | Entity | Model |
| ViewModel | ... | Proxy | Proxy (a.k.a. the X) |
| ... | Update | Model / ViewModel | Update (TBD) |
| View | View | View | View

File: doc/Overview/Mvux/WIP_Overview.md [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+
+### Model
+
+The **Model** that the Update creates (and is wrapped in a generated ViewModel) is typically a simple class or record that contains only data and no logic. For this reason, a record is usually the most appropriate data structure for the Model. C# records are immutable (read-only) and the use of them are encouraged by MVUX. It will be explained further later.
+
+Working with immutable data structures can be a challenging transition for developers who have only previously only worked with `INotifyPropertyChanged` implementing classes. The fundamental difference is that since a record is immutable, making a change to it means creating a new object with different data instead of modifying data within an existing one. Don't worry, the "X" part of the MVUX implementation takes care of this for you.
+
+### Extended
+
+The **eXtended** part of the pattern provides the toolset to adapt between the data-binding engine and immutable data objects, as well as the rest of the functionality that will be covered later on.
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md]: The link should be the advanced pages which is not yet merged. Let's keep this as is for the moment.
I'm updating to "The **eXtended** part of the pattern includes the toolset to adapt between the data-binding engine and immutable data objects, and additional advanced functionality that is outside the scope of this introduction."

File: doc/Overview/Mvux/WIP_Overview.md [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+
+One of the big differences between MVUX and patterns like MVVM is the use of 'Feeds' in place of some of the change-notification properties. While a property has a single value, a feed is similar to a stream of values.
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md]: I don't particularly like the use of quotation marks throughout the page, I prefer changing them all to Feed (with capital F).

File: doc/Overview/Mvux/WIP_Overview.md [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+- It's entirely `async` to ensure a responsive user interface.
+- It is reactive rather than event-driven and so needs less boilerplate code.
+- It encourages the immutability of Model classes, which in turn can lead to improved efficiency and performance, simpler code, and easier object comparison and testing.
+- It automatically tracks and reports the status (or 'state') of asynchronously requested data, which also simplifies the code you need to write. 
+
+Now that you know the fundamentals of the pattern, let's see how to use it to create the simple weather app mentioned above. 
+
+## Using MVUX to create an app
+
+This example creates an `IFeed<WeatherInfo>` that will return the weather, which includes the temperature, loaded asynchronously from a weather service.
```
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md]: Should focus on the user reading the code and understand it on the page or encourage them to create on? This would include detailing all the steps in creating the project using VS etc. etc.

File: doc/Overview/Mvux/WIP_Overview.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md]: How about 'asynchronous connection point'?

File: doc/Overview/Mvux/WIP_Overview.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (nickrandolph) [reply doc/Overview/Mvux/WIP_Overview.md]: At this stage I don't think we can use Model for the entity. As much as it might make sense from the perspective of aligning with MVU, it would require a rework of the way that MVUX works - it generates code for a class called XYZ**Model**, so we should refer to this class in the documentation as **Model**.
For the ViewModel (ie the generated class) we should use **Bindable Proxy**, and this can apply to proxies generated for both model and other entities. We shouldn't refer to Bindable Proxy Model; if we need to be specific then use Bindable Proxy for the Model.
Update can reference a method in the Model but could also be a method on an entity that returns a new instance of the entity (see the Calculator in  SimpleCalculator).
Use **Entity** to refer to data classes (ie the Model in MVVM/MVU)

File: doc/Overview/Mvux/WIP_Overview.md [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+
+Looking at each of the MVUX elements, it's easiest to start with the View.
+
+### View
+
+The **View** is the UI. You can write it with XAML, C#, or a combination of the two, much as you would when using another design pattern.
+
+You use data-binding to bind to the underlying presentation layer (we'll get to it soon) to keep the UI code (the View) separate from the presentation code.
+
+### Update
```
Comment (nickrandolph) [inline doc/Overview/Mvux/WIP_Overview.md:40]: Swtich Model and Update

File: doc/Overview/Mvux/WIP_Overview.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (mrlacey) [reply doc/Overview/Mvux/WIP_Overview.md]: Before making any other changes to this document, I'd work on creating a glossary. Define what words like "entity" mean so that this PR isn't people sharing their opinions and it's documented for future consistency.

[the remark this answers is not in the document]
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md]: We should work on an MVUX glossary page, but that's outside the scope of this PR. I will avoid using 'entity' or other ambiguous words, and if I do I'll clarify its meaning given the context.

File: doc/Overview/Mvux/WIP_Overview.md (hunk 3 above)
Comment (weitzhandler) [reply doc/Overview/Mvux/WIP_Overview.md]: The description has been updated accordingly.

**Review by nickrandolph — CHANGES_REQUESTED** [review]
need to squash edits

**Review by nickrandolph — APPROVED** [review]
